### PR TITLE
Fix Tx.Metadata for normalized transactions

### PR DESCRIPTION
**Pull request by nickchapman-da**
Fix Tx.Metadata for normalized transactions

The primary cause of the bug was the `Tx.Metadata` was referencing node-ids calculated prior to the normalization of the transaction. The bug did *not* manifest if normalization made no change to the node-ids, which was until recently was always the case if the transaction contained no rollback nodes.

With the recent change to `reinterpret` (adding a try-catch handler around the command being reinterpreted), the pre-normalized tx was constructed with transitory node-ids which did not end up matching the final node-ids after normalization... even in the case that the tx contains no rollback nodes!

The fix is to never use the transitory node-ids generated before normalization, specifically those which were in `actionNodeSeeds`, and instead calculate the correct NodeIds in `PartialTransaction.finish`. This is easy, so long as we always collect the actionNodeSeeds in pre-order. Which this PR also ensures.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala [hunk 1]
```diff
@@ -507,6 +507,7 @@ class EngineTest
           lookupPackage,
         )
       isReplayedBy(tx, rtx) shouldBe Right(())
+      txMeta.nodeSeeds shouldBe newMeta.nodeSeeds
```
Comment (remyhaemmerle-da) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala:510]: :+1:

Comment (remyhaemmerle-da) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala:510]: Maybe we could have one more involved test that actually uses rollback node.

**Review by remyhaemmerle-da — APPROVED** [review]
Thanks.

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala (hunk 1 above)
Comment (cocreature) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala:510]: A test for `submit` also would be nice. This really applies to everything as soon as a `try` is involved.

**Review by cocreature — APPROVED** [review]
thanks!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala [hunk 2]
```diff
@@ -320,6 +322,10 @@ private[lf] case class PartialTransaction(
       sb.toString
     }
 
+  private def normalizedNodeSeeds(): NodeSeeds = {
+    ImmArray(actionNodeSeeds.toImmArray.toList.zipWithIndex.map { case (v, i) => (NodeId(i), v) })
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala:326]: You should also avoid using `ImmArray#apply[X](xs: X*)` which is here more for convenience that for efficiency.

The optimal version going in the direction of @cocreature is as follow:
```suggestion
    actionNodeSeeds.toImmArray.toSeq.zipWithIndex.map { case (v, i) => (NodeId(i), v) }.toImmArray
```

**Comment by remyhaemmerle-da** [comment]
I find a bit fragile to use `NodeId` for the node and not for the seed.
We could either: 
- mode the seed directly in the node
- handle everything in preorder traversal (without Node Id) until normalization.

Not for PR though.

**Comment by cocreature** [comment]
> I find a bit fragile to use `NodeId` for the node and not for the seed.
> We could either:
> 
>     * mode the seed directly in the node
> 
>     * handle everything in preorder traversal (without Node Id) until normalization.
> 
> 
> Not for PR though.

yeah I agree it’s a bit messy.

**Comment by oliverse-da** [comment]
Thanks all!

**Pull request merged**